# Auto-dN-dS
A package to automate the retrieval, alignment, and dN/dS calculation of the whole CDS sequences across species.

## Overview of scripts
  1) mastercode.py: main script, with a customizable part at the beginning to set up variables and inputs.

  2) 1list.py: main script, with a customizable part at the beginning to set up variables and inputs.

  3) 2CDS_fetcher.py: main script, with a customizable part at the beginning to set up variables and inputs.

  4) 3align.py: main script, with a customizable part at the beginning to set up variables and inputs.

  5) 4qualityMOS.py: main script, with a customizable part at the beginning to set up variables and inputs.

  6) 5kaks.py: main script, with a customizable part at the beginning to set up variables and inputs.

  7) macse_v2.07.jar: main script, with a customizable part at the beginning to set up variables and inputs.


## Setup
This README file provides instructions on how to set up the project environment by installing the required dependencies for both R and Python.

First of all, download the Github package and navigate in the package directory:

```console
git clone https://github.com/gpmerola/Auto-dN-dS.git
cd Auto-dN-dS
```

Then, install the R and Python dependecies.
To install the required R packages, run the following command:

```console
Rscript requirements.R
```
To install the required Python packages, run the following command:

```console
pip install -r requirements.txt
```

## Usage
  1) Run mastercode.py:

```console
python mastercode.py
```

  2) Provide input regarding which parts of the code to run; type "all" if you want the whole code to run.

      1: list - Performs munging, LD score regression, and prepares SNP files.
      
      2: fetching - Fits the specified structural equation model using the input data.
      
      3: alignemnt - Performs the most computationally intensive step, running a synthethic GWAS on the latent variables specified in the model. Including "r" runs the GWAS in a test mode, using only chromosome 2.
      
      4: mos - Generates Manhattan and QQ plots, performs post-munging, and computes LD score regression for the new phenotype.
      
      5: kaks - Computes genetic correlation between the new phenotype and the input traits.
      
The parts must be run in order, with "3" being the most computationally intensive step.

## Output
The following output files are generated by the script during the analysis workflow. Each file serves a specific purpose in the overall analysis process, from preprocessing and modeling to visualization and correlation analysis. These files are saved in the new working directory created and set at the beginning of the script, provided through the console command.

  1) "dir.txt": Contains the working directory name. It helps share the information between the .R and the .py script.

  2) "LDSC_main.rds": Contains the LDSC output data.

  3) "SNP_files.rds": Contains the sumstats for the SNPs, harmonized.

## License
This project is licensed under the MIT License. See the LICENSE file for details.
